Ensure that a constructor cannot be run twice on the same data account

seanyoung opened the pull request:
The constructor code path was not checking that the data account magic number is not populated.

The change also exposed another issue: when a pure function is called, there should be no load of data magic number, this may result in a access violation.

LucasSte commented:
> The change also exposed another issue: when a pure function is called, there should be no load of data magic number, this may result in a access violation.

This is not solved here, is it?

seanyoung commented:
> > The change also exposed another issue: when a pure function is called, there should be no load of data magic number, this may result in a access violation.
> 
> This is not solved here, is it?

Yes, it is. When I added the magic check for the constructor code path, suddenly I was getting access violations. I've moved the magic load to where it is needed

xermicus reviewed (approved):
I'd expect this change to break some existing codegen tests; I also don't see this fix being verified in a codegen.

seanyoung commented on integration/solana/simple.spec.ts:
```diff
@@ -9,6 +9,11 @@ describe('Simple solang tests', function () {
+        // make sure we can't run the constructor twice
+        await expect(program.methods.new(false)
+            .accounts({ dataAccount: storage.publicKey })
+            .rpc()).rejects.toThrow();
```
Yes, this is formatted by vscode.

seanyoung commented:
> I'd expect this change to break some existing codegen tests; I also don't see this fix being verified in a codegen.

I've added a MockVM test

LucasSte commented on src/codegen/dispatch/solana.rs:
```diff
@@ -445,6 +444,57 @@ fn add_constructor_dispatch_case(
+                loc: Loc::Codegen,
+                ty: Type::Uint(32),
+                value: 0.into(),
+            },
+        },
+    );
+
+    cfg.add(
+        vartab,
+        Instr::BranchCond {
+            cond: Expression::Equal {
+                loc: Loc::Codegen,
+                left: Expression::Variable {
+                    loc: Loc::Codegen,
+                    ty: Type::Uint(32),
+                    var_no: magic,
+                }
+                .into(),
+                right: Expression::NumberLiteral {
+                    loc: Loc::Codegen,
+                    ty: Type::Uint(32),
+                    value: 0.into(),
+                }
+                .into(),
+            },
+            true_block: magic_ok,
+            false_block: magic_bad,
+        },
+    );
+
+    cfg.set_basic_block(magic_bad);
+
+    cfg.add(
+        vartab,
+        Instr::ReturnCode {
+            code: ReturnCode::InvalidDataError,
+        },
+    );
+
+    cfg.set_basic_block(magic_ok);
```
You have here 48 lines of repeated code. What do you think of creating a function?

seanyoung replied:
Good point, fixed

LucasSte reviewed (approved):
The Mac Arm test has failed.
